dotCMS/src/main/java/com/dotcms/rendering/velocity/servlet/VelocityPreviewServlet.java

```diff
@@ -0,0 +1,359 @@
+
+
+        Host host = hostWebAPI.getCurrentHostNoThrow(request);
+
+        PageMode mode = PageMode.get(request);
+        // Checking if host is active
+        boolean hostlive;
+
+
+        try {
+            hostlive = APILocator.getVersionableAPI()
+                .hasLiveVersion(host);
+        } catch (Exception e1) {
+            UtilMethods.closeDbSilently();
+            throw new ServletException(e1);
+        }
+        if (!hostlive) {
+            try {
+                response.sendError(HttpServletResponse.SC_SERVICE_UNAVAILABLE,
+                        LanguageUtil.get("server-unavailable-error-message"));
+            } catch (LanguageException e) {
+                Logger.error(CMSFilter.class, e.getMessage(), e);
+                response.sendError(HttpServletResponse.SC_SERVICE_UNAVAILABLE);
+            }
+            return;
+        }
+
+
+        if (DbConnectionFactory.isMsSql() && LicenseUtil.getLevel() < LicenseLevel.PROFESSIONAL.level) {
+            request.getRequestDispatcher("/portal/no_license.jsp")
+                .forward(request, response);
+            return;
+        }
+
+        if (DbConnectionFactory.isOracle() && LicenseUtil.getLevel() < LicenseLevel.PRIME.level) {
+            request.getRequestDispatcher("/portal/no_license.jsp")
+                .forward(request, response);
+            return;
+        }
+        if (!LicenseUtil.isASAllowed()) {
+            request.getRequestDispatcher("/portal/no_license.jsp")
+                .forward(request, response);
+            return;
+
+        }
+
+        try {
+
+
+            if (uri == null) {
+                response.sendError(500, "VelocityServlet called without running through the CMS Filter");
+                Logger.error(this.getClass(),
+                        "You cannot call the VelocityServlet without passing the requested url via a  requestAttribute called  "
+                                + Constants.CMS_FILTER_URI_OVERRIDE);
+                return;
+            }
+
+
+
+            // we will always need a visitor in admin mode
+            if (mode.isAdmin) {
+                visitorAPI.getVisitor(request, true);
+            }
+
+
+            switch (mode) {
+                case PREVIEW:
+                    Logger.debug(VelocityPreviewServlet.class, "VELOCITY SERVLET I'M ON PREVIEW MODE!!!");
+                    doPreviewMode(request, response);
+                    break;
+                case EDIT:
+                    Logger.debug(VelocityPreviewServlet.class, "VELOCITY SERVLET I'M ON EDIT MODE!!!");
+                    doEditMode(request, response);
+                    break;
+                default:
+                    response.sendError(HttpServletResponse.SC_UNAUTHORIZED);
+                    return;
+            }
+
+
+
+        } catch (ResourceNotFoundException rnfe) {
+            Logger.error(this, "ResourceNotFoundException" + rnfe.toString(), rnfe);
+            response.sendError(404);
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Assign this magic number 404 to a well-named constant, and use the constant instead. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS109)